crml/transaction-payment/src/lib.rs

```diff
@@ -243,48 +241,46 @@ where
 		info: Self::DispatchInfo,
 		len: usize,
 	) -> TransactionValidity {
-		// A cautionary measure to ensure the storage of previous transactions wouldn't extend beyond them.
-		unhashed::kill(&GAS_FEE_EXCHANGE_KEY);
-
-		let tip = self.tip;
-
-		// pay any fees.
-		let fee = Self::compute_fee(len as u32, info, tip);
+		let fee = Self::compute_fee(len as u32, info, self.tip);
 
-		let mut exchange_op = self.fee_exchange.clone();
-
-		// Only mess with balances if fee is not zero.
+		// Only mess with balances if the fee is not zero.
 		if !fee.is_zero() {
-			if let Some(exchange) = exchange_op {
-				let sold_amount = T::BuyFeeAsset::buy_fee_asset(who, fee, &exchange)
-					.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
-				let new_fee_exchange_max = exchange
-					.get_balance()
-					.checked_sub(&sold_amount)
-					.ok_or(TransactionValidityError::from(InvalidTransaction::Payment))?;
-				exchange_op = Some(<TransactionFeeExchange<T>>::new_v1(
-					exchange.get_asset_id(),
-					new_fee_exchange_max,
-				));
+			if let Some(exchange) = &self.fee_exchange {
+				// Buy the CENNZnet fee currency paying with the user's nominated fee currency
+				let fees_purchased = T::BuyFeeAsset::buy_fee_asset(who, fee, &exchange)
+					.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+				// Temporarily store the `FeeExchange` info so that it can be used later by the-
+				// custom `GasHandler` impl see `runtime/src/impls.rs`
+				// This is a hack!
+				storage::unhashed::put(
```

Comment:
This is now saying if the computed fee is zero, we shouldn't save the fee exchange for gas payment. Do we want to couple fee and gas this way?